src/verify.c

```diff
@@ -959,28 +1004,79 @@ enum swupd_code verify_main(void)
 		}
 	}
 
-	ret = add_included_manifests(official_manifest, &subs);
+	/* get the list of subscribed (installed) bundles */
+	read_subscriptions(&all_subs);
+	selected_subs = all_subs;
+	if (cmdline_option_bundles) {
+		info("\n");
+		for (iter = list_head(cmdline_option_bundles); iter; iter = iter->next) {
+			/* let's validate the bundle names provided by the user first */
+			struct file *file;
+			file = search_bundle_in_manifest(official_manifest, iter->data);
+			if (!file) {
+				/* bundle is invalid or no longer available, nothing to do with
+				 * this one for now */
+				invalid_bundle = true;
+				warn("Bundle \"%s\" is invalid or no longer available\n", iter->data);
+				continue;
+			}
+			if (!component_subscribed(all_subs, iter->data)) {
+				/* the requested bundle is not installed, there are two use cases
+				 * where we allow users to specify a bundle that is not installed:
+				 * - when using "verify --fix --bundles" (legacy way to install a bundle)
+				 * - when using "os-install --bundles" or "verify --install --bundles"
+				 * In these cases we need to make sure the subscriptions in bundles_subs
+				 * are also in the all_subs list or they will just be ignored */
+				if ((cmdline_command_verify && cmdline_option_fix) || (cmdline_option_install)) {
+					create_and_append_subscription(&all_subs, iter->data);
+				} else {
+					warn("Bundle \"%s\" is not installed, skipping it...\n", iter->data);
+					continue;
+				}
+			}
+			create_and_append_subscription(&bundles_subs, iter->data);
+		}
+		selected_subs = bundles_subs;
+		/* the user should have specified at least one valid bundle, otherwise finish */
+		if (!bundles_subs) {
+			info("\nNo valid bundles were specified, nothing to be done\n");
+			goto clean_and_exit;
+		}
+	}
+
+	if (cmdline_option_bundles && !cmdline_option_install) {
+		info("\nLimiting %s to the following bundles:\n", cmdline_command_verify ? "verify" : "diagnose", version);
+		for (iter = list_head(bundles_subs); iter; iter = iter->next) {
+			struct sub *sub = iter->data;
+			info(" - %s\n", sub->component);
+		}
+	}
+
+	ret = add_included_manifests(official_manifest, &all_subs);
+	if (cmdline_option_bundles && !(ret & add_sub_ERR)) {
+		ret = add_included_manifests(official_manifest, &bundles_subs);
+	}
 	/* if one or more of the installed bundles were not found in the manifest,
 	 * continue only if --force was used since the bundles could be removed */
-	if (ret) {
-		if (ret & add_sub_BADNAME) {
+	if (ret || invalid_bundle) {
+		if ((ret & add_sub_BADNAME) || invalid_bundle) {
 			if (cmdline_option_force) {
 				if (cmdline_option_picky && cmdline_option_fix) {
-					warn("One or more installed bundles that are not "
+					warn("\nOne or more installed bundles that are not "
 					     "available at version %d will be removed\n",
```

Comment:
We're not removing a bundle if the bundle name is incorrect.

Reply:
when this condition happens it means we found a bundle that was available in an older MoM but is no longer available in the new MoM, it is not really an incorrect bundle name. So this warning is to let the user know that the bundle is going to be removed if he/she --force the update using --picky.

Reply:
btw, I didn't implement that part in this PR, I'm just making the output prettier